File: 백준/Gold/20008. 몬스터를 처치하자！/몬스터를 처치하자！.py
import sys
input = sys.stdin.readline

N, HP = map(int, input().split())
cool = []
dmg = []

for _ in range(N):
    c, d = map(int, input().split())
    cool.append(c)
    dmg.append(d)

answer = float('inf')

def dfs(time, hp, cd):
    global answer
    
    # 몬스터 처치 완료
    if hp <= 0:
        answer = min(answer, time)
        return
    
    # 이미 최적해보다 오래 걸림 → 가지치기
    if time >= answer:
        return
    
    used_skill = False
    
    # 사용 가능한 모든 스킬 시도
    for i in range(N):
        if cd[i] == 0:        # 사용 가능
            used_skill = True

            new_cd = cd[:]    # 쿨타임 복사
            new_hp = hp - dmg[i]

            new_cd[i] = cool[i]  # 스킬 사용 → 쿨타임 설정

            # 1초 흐름 → 모든 쿨타임 감소
            for j in range(N):
                if new_cd[j] > 0:
                    new_cd[j] -= 1

            dfs(time + 1, new_hp, new_cd)

    # 스킬을 아무것도 못 쓰면 1초 기다리기
    if not used_skill:
        new_cd = cd[:]
        for j in range(N):
            if new_cd[j] > 0:
                new_cd[j] -= 1
        dfs(time + 1, hp, new_cd)


# 초기 쿨타임은 모두 0
dfs(0, HP, [0] * N)

print(answer)

File: 백준/Gold/20008. 몬스터를 처치하자！/몬스터를 처치하자！.java
import java.io.*;
import java.util.*;

public class Main{

    static int N;
    static int HP;
    static int[] C;
    static int[] D;
    static int answer = Integer.MAX_VALUE;

    static void dfs(int time, int hp, int[] cd) {
        // 몬스터가 죽으면 종료
        if (hp <= 0) {
            answer = Math.min(answer, time);
            return;
        }

        // 이미 최적해보다 느리면 가지치기
        if (time >= answer) return;

        boolean used = false;

        // 모든 스킬 시도
        for (int i = 0; i < N; i++) {

            if (cd[i] == 0) { // 쓸 수 있는 스킬
                used = true;

                int[] newCd = cd.clone();
                int newHp = hp - D[i];

                newCd[i] = C[i]; // 스킬 쿨타임 초기화

                // 1초 흐름 → 쿨타임 감소
                for (int j = 0; j < N; j++) {
                    if (newCd[j] > 0) newCd[j]--;
                }

                dfs(time + 1, newHp, newCd);
            }
        }

        // 스킬을 하나도 못 쓰면 기다리기
        if (!used) {
            int[] newCd = cd.clone();
            for (int j = 0; j < N; j++) {
                if (newCd[j] > 0) newCd[j]--;
            }
            dfs(time + 1, hp, newCd);
        }
    }
    
    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        HP = Integer.parseInt(st.nextToken());

        C = new int[N];
        D = new int[N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            C[i] = Integer.parseInt(st.nextToken());
            D[i] = Integer.parseInt(st.nextToken());
        }

        dfs(0, HP, new int[N]);

        System.out.println(answer);
    }
}
